sdk/program/src/stake/instruction.rs

```diff
@@ -261,6 +261,27 @@ pub enum StakeInstruction {
     ///   2. `[]` Reference vote account that has voted at least once in the last
     ///      `MINIMUM_DELINQUENT_EPOCHS_FOR_DEACTIVATION` epochs
     DeactivateDelinquent,
+
+    /// Redelegate activated stake to another vote account.
+    ///
+    /// Upon success:
+    ///   * the balance of the delegated stake account will be reduced to the rent exempt minimum,
+    ///     and scheduled for deactivation.
+    ///   * the provided uninitialized stake account will receive the original balance of the
+    ///     delegated stake account, minus the rent exempt minimum, and scheduled for activation to
+    ///     the provided vote account. Any existing lamports in the uninitialized stake account
+    ///     will also be included in the re-delegation.
```

Comment:
Does this totally cover multi-epoch activations?  It'll be fine for the new stake, but what about the previous stake? For example, if I'm on validator A, and redelegate to B, and it takes 4 epochs, then after the first epoch, nothing would be on A, and some on B, so I'd get rewards from B, and none from A.

If there's a slashing event on A during that time, it won't be counted on my stake.  This would allow me to bypass slashes on A since the stake is effectively deactivated immediately.

Reply:
> Does this totally cover multi-epoch activations? 

I'm quite confident this is a yes, but plan to re-read that part of bank.rs again tomorrow to triple check.

> It'll be fine for the new stake, but what about the previous stake? For example, if I'm on validator A, and redelegate to B, and it takes 4 epochs, then after the first epoch, nothing would be on A, and some on B, so I'd get rewards from B, and none from A.

Interestingly since the epoch rewards fully operate on the _contents_ of the stake account and not the _lamports_ of the stake account, everything works out here. During a multi-epoch activation/deactivation, each stake account continues to earn rewards corresponding to the % of their stake that is active at the time.

So while the original stake account is decimated to the rent exempt reserve, it'll still receive at least one final epoch reward before its deactivated by the runtime.

Reply:
Oh right, very clever! In that case, there's a pretty nasty little trick you can do:

* redelegate from validator a to b
* transfer enough lamports into the source account to cover the minimum
* delegate it back to validator a --> the old `redelegate_stake` logic kicks in, and just changes the deactivation epoch but not the stake delegation.

That'll allow you to essentially get infinite rewards if you keep doing it :laughing: so that logic will need to change too.  I'll also need to look at every place where `delegation` is being used to make sure it's always up-to-date.

Reply:
You had me worried there for a second, but unless I'm misunderstanding I don't see how that trick is possible:

During this step:
> delegate it back to validator a --> the old redelegate_stake logic kicks in, and just changes the deactivation epoch but not the stake delegation.

`validate_delegated_amount` reads the actual lamport balance of the source account into `stake_amount`
https://github.com/solana-labs/solana/blob/f14922c2045a13c2aeb41968f0ec2f1fbad3f37b/programs/stake/src/stake_state.rs#L597-L607
and `redelegate` (renamed to  `redelegate_stake` in this PR) resets `delegation.stake` to  `stake_amount` (aka `stake_lamports` in the code ref):
https://github.com/solana-labs/solana/blob/f14922c2045a13c2aeb41968f0ec2f1fbad3f37b/programs/stake/src/stake_state.rs#L125-L127

Reply:
`return` is a bad word: https://github.com/solana-labs/solana/blob/f14922c2045a13c2aeb41968f0ec2f1fbad3f37b/programs/stake/src/stake_state.rs#L115

Reply:
Bah, got it. Looks like the `if` on line 111-112 just needs a `&& stake.delegation.stake == stake_lamports` added to it

Reply:
A less impactful way is to just clear the `voter_pubkey` of the source account after deactivation, so that the source account will always hit this `return` instead:
https://github.com/solana-labs/solana/blob/f14922c2045a13c2aeb41968f0ec2f1fbad3f37b/programs/stake/src/stake_state.rs#L118

Added this in 68d945d5b66b03ecca3e5c20e2fc5833432b9432

Reply:
Nice find btw!

Reply:
I took a slightly different approach to resolving this in `redelegate_stake()` instead, to avoid clearing the `voter_pubkey` as I'm pretty sure doing so will would poorly affect some existing stake movement tracking apps like solana compass